File: experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=ATDD/rep1/test/acceptance/player_queries_test.exs
defmodule BrazilianSoccer.Acceptance.PlayerQueriesTest do
  @moduledoc """
  Category 3: Player Queries. Search by name, nationality, club and position
  through the `search_players` and `get_player` MCP tools.
  """
  use ExUnit.Case, async: true

  alias BrazilianSoccer.Test.MCPClient

  test "look up a single player by name" do
    result = MCPClient.call("get_player", %{"name" => "Neymar"})

    assert result["found"] == true
    assert result["player"]["name"] =~ "Neymar"
    assert result["player"]["nationality"] == "Brazil"
    assert is_integer(result["player"]["overall"])
  end

  test "find all Brazilian players in the dataset" do
    result = MCPClient.call("search_players", %{"nationality" => "Brazil", "limit" => 5000})

    assert result["count"] > 100
    assert Enum.all?(result["players"], &(&1["nationality"] == "Brazil"))
  end

  test "find the highest-rated players, sorted by overall rating" do
    result =
      MCPClient.call("search_players", %{"nationality" => "Brazil", "limit" => 10})

    overalls = Enum.map(result["players"], & &1["overall"])
    assert overalls == Enum.sort(overalls, :desc)
    # The top Brazilian in this FIFA dataset is Neymar.
    assert hd(result["players"])["name"] =~ "Neymar"
  end

  test "filter players by club" do
    result = MCPClient.call("search_players", %{"club" => "Santos", "limit" => 100})

    assert result["count"] > 0
    assert Enum.all?(result["players"], &String.contains?(&1["club"], "Santos"))
  end

  test "filter players by position" do
    result =
      MCPClient.call("search_players", %{
        "club" => "Santos",
        "position" => "GK",
        "limit" => 100
      })

    assert result["count"] > 0
    assert Enum.all?(result["players"], &(&1["position"] == "GK"))
  end

  test "a minimum-rating filter excludes lower-rated players" do
    result = MCPClient.call("search_players", %{"min_overall" => 88, "limit" => 100})

    assert result["count"] > 0
    assert Enum.all?(result["players"], &(&1["overall"] >= 88))
  end

  test "looking up a player who is not in the dataset reports not found" do
    result = MCPClient.call("get_player", %{"name" => "Some Unknown Person 12345"})
    assert result["found"] == false
  end
end

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-8_tooling=none/rep3-failed/test/playerQueries.test.ts
/**
 * ============================================================================
 * Context Block — File: test/playerQueries.test.ts
 * Project: Brazilian Soccer MCP Server
 * Purpose: BDD specs for capability area #3 (Player Queries) over the FIFA
 *          dataset: search by name, nationality, club, position; sorting and
 *          club aggregation.
 * ============================================================================
 */

import { describe, it, expect } from 'vitest';
import { givenLoadedDatabase } from './helpers.js';

describe('Feature: Player Queries', () => {
  it('Scenario: Search a player by name', () => {
    // When I look up Neymar
    const db = givenLoadedDatabase();
    const results = db.searchPlayers({ name: 'Neymar' });
    // Then I find the Brazilian forward with a high rating
    expect(results.length).toBeGreaterThan(0);
    const top = results[0];
    expect(top.nationality).toBe('Brazil');
    expect(top.overall).toBeGreaterThan(85);
  });

  it('Scenario: Find all Brazilian players in the dataset', () => {
    const db = givenLoadedDatabase();
    const brazilians = db.searchPlayers({ nationality: 'Brazil', limit: 1000 });
    expect(brazilians.length).toBeGreaterThan(500);
    for (const p of brazilians.slice(0, 50)) expect(p.nationality).toBe('Brazil');
  });

  it('Scenario: Top Brazilian players are sorted by overall rating', () => {
    const db = givenLoadedDatabase();
    const top = db.searchPlayers({ nationality: 'Brazil', sortBy: 'overall', limit: 5 });
    expect(top[0].name.toLowerCase()).toContain('neymar');
    for (let i = 1; i < top.length; i++) {
      expect(top[i - 1].overall ?? 0).toBeGreaterThanOrEqual(top[i].overall ?? 0);
    }
  });

  it('Scenario: Filter players by club', () => {
    const db = givenLoadedDatabase();
    const santos = db.searchPlayers({ club: 'Santos' });
    expect(santos.length).toBeGreaterThan(0);
    for (const p of santos) expect(p.clubKey).toContain('santos');
  });

  it('Scenario: Filter players by position', () => {
    const db = givenLoadedDatabase();
    const keepers = db.searchPlayers({ position: 'GK', nationality: 'Brazil', limit: 20 });
    expect(keepers.length).toBeGreaterThan(0);
    for (const p of keepers) expect(p.position).toBe('GK');
  });

  it('Scenario: Aggregate Brazilian players by club', () => {
    const db = givenLoadedDatabase();
    const byClub = db.playersByClub({ nationality: 'Brazil', minPlayers: 5 });
    expect(byClub.length).toBeGreaterThan(0);
    const first = byClub[0];
    expect(first.players).toBeGreaterThanOrEqual(5);
    expect(first.avgRating).toBeGreaterThan(0);
  });
});
